### PR TITLE
Enable CI lint gh action on ROCm

### PR DESCRIPTION
**Pull request by VeeraRajasekhar**
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/common/ck_fused_attn/src/ck_fused_attn_utils.cpp [hunk 1]
```diff
@@ -76,7 +76,13 @@ void set_aiter_asm_dir() {
...
   std::call_once(aiter_asm_dir_once, []() {
     Dl_info info;
-    dladdr((void*)set_aiter_asm_dir, &info);
+    // dladdr expects void*; avoid reinterpret_cast<void*>(fn) (not ISO C++).
+    union {
+      void (*fn)();
+      void *addr;
+    } sym{};
+    sym.fn = set_aiter_asm_dir;
+    dladdr(sym.addr, &info);
```
Comment (Micky774) [inline transformer_engine/common/ck_fused_attn/src/ck_fused_attn_utils.cpp:85]: IMO this is unnecessary. Yes, it quiets the warning, but the warning is irrelevant for us granted our support is POSIX focused to begin with. From the [`dlopen` man page](https://man7.org/linux/man-pages/man3/dlopen.3.html):
```
           /* According to the ISO C standard, casting between function
              pointers and 'void *', as done above, produces undefined results.
              POSIX.1-2001 and POSIX.1-2008 accepted this state of affairs and
              proposed the following workaround:

                  *(void **) &cosine = dlsym(handle, "cos");

              This (clumsy) cast conforms with the ISO C standard and will
              avoid any compiler warnings.

              The 2013 Technical Corrigendum 1 to POSIX.1-2008 improved matters
              by requiring that conforming implementations support casting
              'void *' to a function pointer.  Nevertheless, some compilers
              (e.g., gcc with the '-pedantic' option) may complain about the
              cast used in this program.  */
```
the union trick here provides no additional safety -- it's still undefined behavior technically speaking -- and will break in the same circumstances (non-POSIX risk).

All things considered, I'd rather we keep things as-is, and if we really want to deal with the warning, we can make a small utility to use pragmas to suppress the warnings locally around the cast.

File: transformer_engine/common/gemm/rocm_gemm.cu [hunk 2]
```diff
@@ -117,7 +117,7 @@ public:
     data[key][stream] = item; 
   }
 
-  ObjCache(void (*a_offload)(const Data&)): offload(a_offload) {}
+  explicit ObjCache(void (*a_offload)(const Data&)): offload(a_offload) {}
```
Comment (Micky774) [inline transformer_engine/common/gemm/rocm_gemm.cu:120]: Do we really want these constructors to be explicit? Are they even used implicitly anywhere in our codebase?

File: commit.txt
Comment (Micky774) [inline commit.txt:1]: I'm assuming this won't be part of the final PR?

**Review by Micky774 — COMMENTED** [review]
Does this cause divergence from upstream's source code in inherited files? Also, why do we need linting as a CI action instead of e.g. a pre-commit?

File: commit.txt
Comment (VeeraRajasekhar) [reply commit.txt:1]: Yes, This is for me to keep track of the issues fixed, will remove this.

File: transformer_engine/common/ck_fused_attn/src/ck_fused_attn_utils.cpp [hunk 3]
```diff
@@ -130,9 +136,10 @@ std::ostream* get_ck_log_stream() {
       if (!log_dir_str.empty() && log_dir_str != "0") {
         if (log_dir_str == "1") {
           log_stream = &std::cout;
-        }
-        else if (open_ck_fused_attn_log_file(log_file, "ck_fused_attn", log_dir_str)) {
```
Comment (ipanfilo) [inline transformer_engine/common/ck_fused_attn/src/ck_fused_attn_utils.cpp:134]: What is a warning for if-else if? I think it is used a lot in our code

File: transformer_engine/common/gemm/rocm_gemm.cu [hunk 4]
```diff
@@ -769,14 +769,17 @@ protected:
       }
 
 #if HIPBLASLT_VERSION_MAJOR > 0 || HIPBLASLT_VERSION_MINOR >= 15
-      if (cfg.scaling_mode < 0 || cfg.scaling_mode >= (int)HIPBLASLT_MATMUL_MATRIX_SCALE_END)
```
Comment (ipanfilo) [inline transformer_engine/common/gemm/rocm_gemm.cu:772]: Line length and { at the end of the line are understood but lint does not require duplicate body of If.

File: transformer_engine/common/normalization/common.cpp [hunk 5]
```diff
@@ -544,24 +544,26 @@ NormalizationPlanBase* NormalizationPlanRegistry::getNormalizationPlan(
     plan = std::make_unique<CudnnNormalizationPlan>(NormType, NormStage, wtype, itype, otype, ctype,
                                                     batch_size, hidden_size, sm_count,
                                                     zero_centered_gamma, mode, training);
-  } else
+  }
 #endif
```
Comment (ipanfilo) [inline transformer_engine/common/normalization/common.cpp]: Let's get rid of nested ifs. If splitting else-if does not work, better add dummy 'if (false) {' for ROCm instead of 'if (NOrmBAckend...'

File: transformer_engine/common/__init__.py [hunk 6]
```diff
@@ -51,8 +51,8 @@ def _is_package_installed_from_wheel(package) -> bool:
     if not te_wheel_file:
         return False
 
-    with te_wheel_file.open("r") as f:
-        for line in f:
```
Comment (ipanfilo) [inline transformer_engine/common/__init__.py:55]: It is upstream code. Why change it?

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 7]
```diff
@@ -9,9 +9,10 @@
 import warnings
 from dataclasses import dataclass, replace
 from functools import partial, reduce
-from packaging import version
 from typing import Optional, Tuple
 
+from packaging import version
```
Comment (ipanfilo) [inline transformer_engine/jax/cpp_extensions/attention.py]: Why is it needed?

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 8]
```diff
@@ -8,16 +8,18 @@
...
 import jax
 import jax.numpy as jnp
 from jax import dtypes, ffi
-from jax.experimental.custom_partitioning import SdyShardingRule, BATCHING
+if version.parse(jax.__version__) >= version.parse("0.5.0"):
+    from jax.experimental.custom_partitioning import SdyShardingRule
+from jax.experimental.custom_partitioning import BATCHING
 from jax.interpreters.mlir import ir
 from jax.sharding import PartitionSpec
-from .misc import is_hip_extension
```
Comment (ipanfilo) [inline transformer_engine/jax/cpp_extensions/normalization.py:20]: Does the file have ROCm code?

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 9]
```diff
@@ -8,16 +8,18 @@
...
-from packaging import version
 from typing import Optional, Union
 
+from packaging import version
+
 import jax
 import jax.numpy as jnp
 from jax import dtypes, ffi
-from jax.experimental.custom_partitioning import SdyShardingRule, BATCHING
+if version.parse(jax.__version__) >= version.parse("0.5.0"):
```
Comment (ipanfilo) [inline transformer_engine/jax/cpp_extensions/normalization.py]: IT is not needed, we don;t actually support JAX < 0.5

File: transformer_engine/jax/quantize/helper.py [hunk 10]
```diff
@@ -26,14 +26,8 @@
 import jax.numpy as jnp
 from flax.core.frozen_dict import FrozenDict
 
-from ..util import is_hip_extension, get_jnp_float8_e4m3_type, get_jnp_float8_e5m2_type
+import transformer_engine_jax as tejax
```
Comment (ipanfilo) [inline transformer_engine/jax/quantize/helper.py]: All those changes will result in IFU conflicts

File: transformer_engine/jax/setup.py [hunk 11]
```diff
@@ -51,11 +51,11 @@
 from build_tools.te_version import te_version
 from build_tools.jax import setup_jax_extension, install_requirements, test_requirements
 
-from pybind11.setup_helpers import build_ext as BuildExtension
```
Comment (ipanfilo) [inline transformer_engine/jax/setup.py:54]: Why is it needed?

**Review by ipanfilo — CHANGES_REQUESTED** [review]

**Review by Micky774 — CHANGES_REQUESTED** [review]
@VeeraRajasekhar I still have questions from my initial review unanswered.

The main one which @ipanfilo has also asked is: why such divergence? It will be a huge maintenance burden to reconcile against _every_ subsequent IFU, unless NV upstream implements these _exact same_ decisions (afaik they don't?).

Enforcing linting on ROCm-only files is fine, but we have constraints regarding minimizing diff with upstream even if it means letting some things go "un-linted".

File: transformer_engine/common/gemm/rocm_gemm.cu (hunk 2 above)
Comment (VeeraRajasekhar) [reply transformer_engine/common/gemm/rocm_gemm.cu:120]: they aren’t used implicitly anywhere; the only constructions are direct ObjCache<T,K>(nullptr) from ObjPool and direct init of service_stream_cache with a lambda. explicit was added only to satisfy cpplint runtime/explicit. If we prefer not to mark these callbacks as explicit, we can drop explicit and suppress that line with NOLINT for cpplint instead.

There are no APIs that take an ObjCache by value. So explicit is not required for correctness, only for style / tooling.

File: transformer_engine/common/ck_fused_attn/src/ck_fused_attn_utils.cpp (hunk 3 above)
Comment (VeeraRajasekhar) [reply transformer_engine/common/ck_fused_attn/src/ck_fused_attn_utils.cpp:134]: The category was readability/braces. The message was along the lines of: “If an else has a brace on one side, it should have it on both.”

So the warning showed up because cpplint’s readability/braces heuristic fired on this if / else if layout, not because else if is forbidden. Nesting as else { if (...) { ... } } makes the structure obvious to the linter and cleared the warning.

File: transformer_engine/common/gemm/rocm_gemm.cu (hunk 4 above)
Comment (VeeraRajasekhar) [reply transformer_engine/common/gemm/rocm_gemm.cu:772]: Yes, the duplicate was to fix the readability/braces warning, I have restructured it to compute a bool in preprocessor-only branches, then use one if body.

File: transformer_engine/common/normalization/common.cpp (hunk 5 above)
Comment (VeeraRajasekhar) [reply transformer_engine/common/normalization/common.cpp]: Dropped the nested if (!plan) / inner if (Forward) structure. and added the following more clear structure:
> CUDA: if (Cudnn) … else if (Forward TE) … else (backward TE) in one #ifndef __HIP_PLATFORM_AMD__ block.
> ROCm: if (Forward TE) … else (backward TE) in #else, with mode/training on TE constructors.

File: transformer_engine/jax/cpp_extensions/attention.py (hunk 7 above)
Comment (VeeraRajasekhar) [reply transformer_engine/jax/cpp_extensions/attention.py]: This is used to check the following

```
if version.parse(jax.__version__) >= version.parse("0.5.0"):
    from jax.experimental.custom_partitioning import SdyShardingRule
```

I will update the codebase to remove all these checks as we no longer support Jax<0.5.0 as mentioned in https://github.com/ROCm/TransformerEngine/pull/547#discussion_r3121416919

File: transformer_engine/jax/cpp_extensions/normalization.py (hunk 8 above)
Comment (VeeraRajasekhar) [reply transformer_engine/jax/cpp_extensions/normalization.py:20]: Yes, jax version checks were added.

File: transformer_engine/jax/cpp_extensions/normalization.py (hunk 9 above)
Comment (VeeraRajasekhar) [reply transformer_engine/jax/cpp_extensions/normalization.py]: Updated

File: transformer_engine/jax/setup.py (hunk 11 above)
Comment (VeeraRajasekhar) [reply transformer_engine/jax/setup.py:54]: Reverted and added # pylint: disable=ungrouped-imports

File: transformer_engine/pytorch/triton_kernels/gmm/gmm_kernels.py [hunk 12]
```diff
@@ -255,7 +256,7 @@ def gmm_kernel(
         tl.device_assert(num_tiles >= 0, "num_tiles < 0")
 
         # Loop through tiles of current MM problem.
-        while tile >= last_mm_tile and tile < last_mm_tile + num_tiles:
+        while last_mm_tile <= tile < last_mm_tile + num_tiles:
```
Comment (Micky774) [inline transformer_engine/pytorch/triton_kernels/gmm/gmm_kernels.py:259]: Triton doesn't support this multi-comparison -- it looks like the linter is trying to enforce python semantics on it

File: transformer_engine/common/ck_fused_attn/src/ck_fused_attn_utils.cpp (hunk 3 above)
Comment (ipanfilo) [reply transformer_engine/common/ck_fused_attn/src/ck_fused_attn_utils.cpp:134]: Move them to single line then but do not create nested ifs

File: transformer_engine/common/gemm/rocm_gemm.cu (hunk 4 above)
Comment (ipanfilo) [reply transformer_engine/common/gemm/rocm_gemm.cu:772]: It looks like there is discrepancy here. .clang-format sets line width 100 and this is what IDE uses. So cpplint should be configured accorfingly

**Review by ipanfilo — CHANGES_REQUESTED** [review]
